# Introduction

In this Session, you'll learn all about **[pandas](https://pandas.pydata.org)**, the most popular Python library for data analysis.

Along the way, you'll complete several hands-on exercises with real-world data.  We recommend that you work on the exercises while reading the corresponding tutorials.  

**To start the first exercise, please click [here](https://www.google.com).**

In this tutorial, you will learn how to create your own data, along with how to work with data that already exists.  

# Getting started

To use pandas, you'll typically start with the following line of code.

In [1]:
import pandas as pd

# Creating data

There are two core objects in pandas: the **DataFrame** and the **Series**.

### DataFrame

A DataFrame is a table. It contains an array of individual *entries*, each of which has a certain *value*. Each entry corresponds to a row (or *record*) and a *column*.

For example, consider the following simple DataFrame:

In [2]:
pd.DataFrame({'Yes': [50, 21], 'No': [131, 2]})

Yes   No
0   50  131
1   21    2

In this example, the "0, No" entry has the value of 131. The "0, Yes" entry has a value of 50, and so on.

DataFrame entries are not limited to **integers**. For instance, here's a DataFrame whose values are **strings**:

In [3]:
pd.DataFrame({'Bob': ['I liked it.', 'It was awful.'], 'Sue': ['Pretty good.', 'Bland.']})

Bob           Sue
0    I liked it.  Pretty good.
1  It was awful.        Bland.

We are using the `pd.DataFrame()` constructor to generate these DataFrame objects. The syntax for declaring a new one is a dictionary whose keys are the column names (`Bob` and `Sue` in this example), and whose values are a list of entries. This is the standard way of constructing a new DataFrame, and the one you are most likely to encounter.

The dictionary-list constructor assigns values to the *column labels*, but just uses an ascending count from 0 (0, 1, 2, 3, ...) for the *row labels*. Sometimes this is OK, but oftentimes we will want to assign these labels ourselves.

The list of row labels used in a DataFrame is known as an **Index**. We can assign values to it by using an `index` parameter in our constructor:

In [4]:
pd.DataFrame({'Bob': ['I liked it.', 'It was awful.'], 
              'Sue': ['Pretty good.', 'Bland.']},
             index=['Product A', 'Product B'])

Bob           Sue
Product A    I liked it.  Pretty good.
Product B  It was awful.        Bland.

### Series

A Series, by contrast, is a sequence of data values. If a DataFrame is a table, a Series is a list. And in fact you can create one with nothing more than a list:

In [5]:
pd.Series([1, 2, 3, 4, 5])

0    1
1    2
2    3
3    4
4    5
dtype: int64

A Series is, in essence, a single column of a DataFrame. So you can assign column values to the Series the same way as before, using an `index` parameter. However, a Series does not have a column name, it only has one overall `name`:

In [6]:
pd.Series([30, 35, 40], index=['2015 Sales', '2016 Sales', '2017 Sales'], name='Product A')

2015 Sales    30
2016 Sales    35
2017 Sales    40
Name: Product A, dtype: int64

The Series and the DataFrame are intimately related. It's helpful to think of a DataFrame as actually being just a bunch of Series "glued together". We'll see more of this in the next section of this tutorial.

# Reading data files

Being able to create a DataFrame or Series by hand is handy. But, most of the time, we won't actually be creating our own data by hand. Instead, we'll be working with data that already exists.

Data can be stored in any of a number of different forms and formats. By far the most basic of these is the humble CSV file. When you open a CSV file you get something that looks like this:

```
Product A,Product B,Product C,
30,21,9,
35,34,1,
41,11,11
```

So a CSV file is a table of values separated by commas. Hence the name: "Comma-Separated Values", or CSV.

Let's now set aside our toy datasets and see what a real dataset looks like when we read it into a DataFrame. We'll use the `pd.read_csv()` function to read the data into a DataFrame. This goes thusly:

In [17]:
df = pd.read_csv("PakistanDroneAttacksWithTemp Ver 9 (October 19, 2017).csv", encoding='cp1252')

We can use the `shape` attribute to check how large the resulting DataFrame is:

In [18]:
df.shape

(405, 25)

So our new DataFrame has 405 records split across 25 different columns. That's almost 10125 entries!

We can examine the contents of the resultant DataFrame using the `head()` command, which grabs the first five rows:

In [19]:
df.head()

S#                         Date   Time  \
0  1.0        Friday, June 18, 2004  22:00   
1  2.0         Sunday, May 08, 2005  23:30   
2  3.0  Thursday, December 01, 2005    NaN   
3  4.0     Friday, January 06, 2006    NaN   
4  5.0     Friday, January 13, 2006   3:00   

                               Location              City Province  \
0                             Near Wana  south Waziristan     FATA   
1          Mir Ali (Near Afghan Border)  North Waziristan     FATA   
2                   Haisori- Miran Shah  North Waziristan     FATA   
3  Saidgai village- 115km north of Wana  North Waziristan     FATA   
4                      Damadola Village    Bajaur  Agency     FATA   

   No of Strike  Al-Qaeda  Taliban  Civilians Min  ...  Injured Min  \
0           1.0       NaN      1.0            0.0  ...          NaN   
1           1.0       1.0      NaN            0.0  ...          NaN   
2           1.0       1.0      NaN            0.0  ...          NaN   
3           1.0       NaN      NaN            NaN  ...          NaN   
4           1.0       NaN      NaN            0.0  ...          NaN   

   Injured Max  Women/Children    \
0          NaN                 N   
1          NaN                 N   
2          2.0               NaN   
3          2.0               NaN   
4          2.0                 Y   

                              Special Mention (Site)  \
0  Blast occured in courtyard of the house of lon...   
1  Drone struck a car driven by local warlord- ki...   
2                  Explosive occurred at a mud house   
3                                                NaN   
4  Three houses were tarheted in Damadola village...   

                                            Comments  \
0                                    Village in Wana   
1  Civilian killied was Samiullah Khan who was a ...   
2  No. 3 Al-Qaeda's Leader AbuHamza Rabia killed ...   
3                                                NaN   
4  Masood Khan house was among those bombed. Want...   

                                          References  Longitude Latitude  \
0       http://archives.dawn.com/2004/06/19/top1.htm    69.9000  33.0333   
1               http://www.msnbc.msn.com/id/7847008/    70.1455  32.9746   
2  http://edition.cnn.com/2005/WORLD/asiapcf/12/0...    70.1455  32.9746   
3  http://www.reuters.com/article/2007/04/27/us-p...    70.1455  32.9746   
4  http://www.dailytimes.com.pk/default.asp?page=...    71.5000  34.6833   

  Temperature(C) Temperature(F)  
0         28.475         83.255  
1         11.475         52.655  
2          7.080         44.744  
3          0.535         32.963  
4         10.025         50.045  

[5 rows x 25 columns]